router2023/main.md

```diff
@@ -0,0 +1,454 @@
+It has 5 ethernet ports, it supports Wifi 6, and can be equipped with a nvme disk - perfect!
+
+At this moment I had literally zero knowledge about embedded systems and ARM devices, all I knew was that you could install there OpenWRT, and since OpenWRT is a linux
+I figured out that installing NixOS wouldn't be much of a problem.
+
+I also found some unofficial [arch](https://forum.banana-pi.org/t/bpi-r3-imagebuilder-r3-archlinux-kernel-v6-3/15089),
+[ubuntu](https://forum.banana-pi.org/t/bpi-r3-ubuntu-22-04-image/14956) and [debian](https://forum.banana-pi.org/t/bpi-r3-debian-bullseye-image/14541)
+images linked in the [bpir3 documentation](https://wiki.banana-pi.org/Banana_Pi_BPI-R3) which only convinced me that this would be fairly easy.
+Well, I was wrong. I mean, in the end, it turned out not to be that hard provided that one knows what they were doing, but, let me remind you, I didn't.
+Tbh there are still a lot of things that I don't understand but I learned a ton and I wanted to share that.
+
+If you find anything incorrect in this article please don't hesitate to suggest a correction.
+
+## Preface
+
+This is not a tutorial on how to install NixOS on some ARM based board. This just me writing about what I have learned while doing that.
+It might have some educational value but there are better resources, especially the great and always evolving nix wiki - https://nixos.wiki/wiki/NixOS_on_ARM
+
+## Is your device supported?
+
+Before we even begin to attempt installing our beloved operating system on the device we should first asses whether it is supported by the mainline kernel or not.
+
+What does that it mean to be supported?
+
+The kernel needs to know what kind of hardware components are available to it. This is information is provided by so called device-tree.
+The device-tree is a tree of descriptions of every device supported by linux kernel, and it lives in the linux repository.
+In our case bpir3 is a device and there should be a device-tree entry for it. This entry will then list all its components together with their physical
+addresses and drivers that are required to communicate with them.
+
+Here is a part of it that defines two buttons (reset and wps):
+
+```
+	keys {
+		compatible = "gpio-keys";
+
+		factory {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&pio 9 GPIO_ACTIVE_LOW>;
+		};
+
+		wps {
+			label = "wps";
+			linux,code = <KEY_WPS_BUTTON>;
+			gpios = <&pio 10 GPIO_ACTIVE_LOW>;
+		};
+	};
+```
+
+Device-tree files (dts) get compiled via device-tree compiler (dtc) into a device-tree blob (dtb) that then gets loaded by the bootloader and/or kernel.
+(There are also dtsi files that are used to extract common definitions e.g. SoC. The dtsi files are meant to be included by regular dts files.)
+
+Luckily, we don't have to always recompile the whole device-tree whenever we want to test some new device.
```

Comment:
Compiling a whole device tree is fast, so I'm not sure that's the reason overlays exist. I think (amateur guess here) it's more to do with being able to use the library of device trees in the kernel tree without having to fork the kernel to apply a little modification.

Reply:
Sounds reasonable, I will change it that way, explicitly saying that it might not be 100% true :+1: